### PR TITLE
CP-1604 Functional Coverage Working For Real I Promise

### PR DESCRIPTION
**Pull request by christopherdaly-wf**
@evanweible-wf @jayudey-wf

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,13 +1,20 @@
 language: dart
+#sudo: required
```
Comment (evanweible-wf) [inline .travis.yml]: Get rid of this if it isn't needed

File: lib/src/task_process.dart [hunk 2]
```diff
@@ -60,4 +59,40 @@ class TaskProcess {
...
+      args.add(pid.toString());
+      var pgrep = new TaskProcess(executable, args);
+      pgrep.stdout.listen((l) {
+        cpids.add(int.parse(l));
+      });
+      pgreps.add(pgrep);
+    }
+    for (int i = 0; i < pgreps.length; i++) {
+      await pgreps[i].done;
+    }
```
Comment (evanweible-wf) [inline lib/src/task_process.dart]:
```dart
await new Future.wait(pgreps.map((pgrep) => pgrep.done));
```

File: lib/src/tasks/coverage/api.dart [hunk 3]
```diff
@@ -213,41 +215,50 @@ class CoverageTask extends Task {
...
-      String executable = 'pub';
-      List args = [
-        'run',
-        'coverage:collect_coverage',
-        '--port=${observatoryPort}',
-        '-o',
-        collection.path
-      ];
+//      for (int j = 0; j < observatoryPorts.length; j++) {
+      for (int j = observatoryPorts.length - 1; j >= 0; j--) {
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Why are you iterating over these ports in reverse? Might be worth adding a comment.
Also get rid of the commented out line.

File: lib/src/tasks/coverage/api.dart [hunk 4]
```diff
@@ -317,8 +331,14 @@ class CoverageTask extends Task {
 
     Map mergedJson = JSON.decode(collections.first.readAsStringSync());
     for (int i = 1; i < collections.length; i++) {
-      Map coverageJson = JSON.decode(collections[i].readAsStringSync());
-      mergedJson['coverage'].addAll(coverageJson['coverage']);
+      if (!collections[i].existsSync()) continue;
+      String coverage = collections[i].readAsStringSync();
+      if (coverage.isEmpty) {
+//        mergedJson['coverage'].addAll({});
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: If this isn't needed, get rid of this if block altogether and just use `if (coverage.isNotEmpty) {`

File: lib/src/tasks/coverage/api.dart [hunk 5]
```diff
@@ -331,6 +351,8 @@ class CoverageTask extends Task {
     return coverage;
   }
 
+  List<int> observatoryPort;
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Since this is a list now, it should be `observatoryPorts`

File: lib/src/tools/selenium.dart [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:uuid/uuid.dart';
+
+import 'package:dart_dev/src/task_process.dart';
+import 'package:dart_dev/src/util.dart';
+
+/// An installer and runner for a standalone Selenium server.
```
Comment (evanweible-wf) [inline lib/src/tools/selenium.dart]: I think I originally intended for this Helper to install selenium, but that's no longer the case. Should remove "installer" from this comment.

File: test/fixtures/copy_license/no_licenses/lib/non-utf8 [hunk 7]
```diff
@@ -1 +0,0 @@
-�
```
Comment (evanweible-wf) [inline test/fixtures/copy_license/no_licenses/lib/non-utf8]: Was this file actually removed? Or is GitHub just not picking up the `test/fixtures` -> `test_fixtures` rename?

File: test/integration/coverage_test.dart [hunk 8]
```diff
@@ -38,6 +41,10 @@ Future<bool> runCoverage(String projectPath, {bool html: false}) async {
   }
 
   List args = ['run', 'dart_dev', 'coverage'];
+  if (functional) {
+    args.add('--functional');
+    args.add('--no-unit');
```
Comment (evanweible-wf) [inline test/integration/coverage_test.dart]: Why should including the functional test suite prevent us from running unit tests?

Comment (evanweible-wf) [inline test/integration/coverage_test.dart]: Oh I didn't realize this was in the test setup. This is probably fine.

File: test/integration/coverage_test.dart [hunk 9]
```diff
@@ -83,6 +90,14 @@ void main() {
       expect(lcov.existsSync(), isTrue);
     }, timeout: new Timeout(new Duration(seconds: 90)));
 
+    test('should generate coverage for Functional tests', () async {
+      expect(await runCoverage(projectWithFunctionalTests, functional: true),
+          isTrue);
+      File lcov =
+          new File('$projectWithFunctionalTests/coverage/coverage.lcov');
+      expect(lcov.existsSync(), isTrue);
+    }, timeout: new Timeout(new Duration(seconds: 300)));
```
Comment (evanweible-wf) [inline test/integration/coverage_test.dart]: I really need to find a way to speed up testing in this repo.

File: test/integration/examples_test.dart [hunk 10]
```diff
@@ -33,9 +33,10 @@ Future<bool> serveExamplesFor(String projectPath) async {
   bool served = false;
 
   Pattern pubServePattern = new RegExp(r'pub serve .* example');
-  process.stdout.listen((line) {
+  process.stdout.listen((line) async {
     if (line.contains(pubServePattern)) {
       served = true;
+      await process.killAllChildren();
```
Comment (evanweible-wf) [inline test/integration/examples_test.dart]: Why is this necessary?

File: lib/src/tasks/coverage/api.dart [hunk 11]
```diff
@@ -332,12 +332,8 @@ class CoverageTask extends Task {
...
       if (!collections[i].existsSync()) continue;
       String coverage = collections[i].readAsStringSync();
-      if (coverage.isEmpty) {
-//        mergedJson['coverage'].addAll({});
-      } else {
-        Map coverageJson = JSON.decode(collections[i].readAsStringSync());
-        mergedJson['coverage'].addAll(coverageJson['coverage']);
-      }
+      Map coverageJson = JSON.decode(collections[i].readAsStringSync());
+      mergedJson['coverage'].addAll(coverageJson['coverage']);
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: We still want the `isEmpty` check here to see if the file was completely empty. Something like this:

```dart
String coverage = collections[i].readAsStringSync();
if (coverage.isNotEmpty) {
  Map coverageJson = JSON.decode(coverage);
  mergedJson['coverage'].addAll(coverageJson['coverage']);
}
```

Comment (christopherdaly-wf) [inline lib/src/tasks/coverage/api.dart]: So we do not want to fail if a coverage file is completely empty?

Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: I'm not actually sure on that one.. I seem to remember that an empty coverage file may just mean that there was no coverage to be collected, which could be the case since we get several observatory ports but only one or a subset of them actually have coverage to be collected.

File: lib/src/tasks/coverage/api.dart [hunk 12]
```diff
@@ -332,8 +332,10 @@ class CoverageTask extends Task {
     for (int i = 1; i < collections.length; i++) {
       if (!collections[i].existsSync()) continue;
       String coverage = collections[i].readAsStringSync();
-      Map coverageJson = JSON.decode(collections[i].readAsStringSync());
-      mergedJson['coverage'].addAll(coverageJson['coverage']);
+      if (coverage.isNotEmpty) {
+        Map coverageJson = JSON.decode(collections[i].readAsStringSync());
```
Comment (evanweible-wf) [inline lib/src/tasks/coverage/api.dart]: Use `JSON.decode(coverage)` here instead of reading the file again.

**Comment by evanweible-wf** [comment]
@christopherdaly-wf I tested this locally, albeit minimally, but seems to work well! We should get some other reviewers, but I'd say this is looking good. Also don't worry about the coverage CI check - I'll lower the threshold. Some day I'll get around to improving the testing story for this repo.

@trentgrover-wf @maxwellpeterson-wf @dustinlessard-wf @jayudey-wf

**Comment by trentgrover-wf** [comment]
+1

**Comment by maxwellpeterson-wf** [comment]
+1

**Comment by evanweible-wf** [comment]
@jayudey-wf ready for merge

**Comment by jayudey-wf** [comment]
found some issues in testing, following up with Chris

**Comment by jayudey-wf** [comment]
QA Resource Approval: +10

- [ ] Testing instruction (derived from Chris)
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - verified that functional test coverage was working as expected and that instructions found within the readme were sufficient to allow for a user to set it up
- [ ] Unit test created/updated
- [ ] All unit tests pass 

Merging into master.
